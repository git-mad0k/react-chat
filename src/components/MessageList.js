import React from 'react';
import PropTypes from 'prop-types';
import { withStyles } from '@material-ui/core/styles';
import { withRouter } from 'react-router-dom';

import Typography from '@material-ui/core/Typography';
import Paper from '@material-ui/core/Paper';
import MessageItem from './MessageItem';

const styles = theme => ({
  messagesWrapper: {
    paddingTop: '56px',
    overflowX: 'auto',
    paddingBottom: '130px',
    width: '100%',
    height: '100%',
  },

  paper: {
    padding: theme.spacing.unit * 3,
  },
});

class MessageList extends React.Component {
  static propTypes = {
    classes: PropTypes.objectOf(PropTypes.string).isRequired,
    messages: PropTypes.arrayOf(PropTypes.shape({
      chatId: PropTypes.string.isRequired,
      createdAt: PropTypes.string.isRequired,
      content: PropTypes.string,
      sender: PropTypes.shape({
        firstName: PropTypes.string.isRequired,
        lastName: PropTypes.string.isRequired,
        username: PropTypes.string.isRequired,
        _id: PropTypes.string.isRequired,
      }),
      statusMessage: PropTypes.bool,
      updatedAt: PropTypes.string,
      _id: PropTypes.string.isRequired,
    })),
    match: PropTypes.shape({
      params: PropTypes.object.isRequired,
    }).isRequired,
    activeUser: PropTypes.shape({
      firstName: PropTypes.string,
      lastName: PropTypes.string,
      username: PropTypes.string,
      isMember: PropTypes.bool.isRequired,
      isCreator: PropTypes.bool.isRequired,
      isChatMember: PropTypes.bool.isRequired,
    }).isRequired,
  };
  static defaultProps = {
    messages: [],
  };
  componentDidMount() {
    this.scrollDownHistory();
  }

  componentDidUpdate() {
    this.scrollDownHistory();
  }

  scrollDownHistory() {
    if (this.messagesWrapper) {
      this.messagesWrapper.scrollTop = this.messagesWrapper.scrollHeight;
    }
  }

  render() {
    const {
      classes, messages, match, activeUser,
    } = this.props;
    if (!match.params.chatId) {
      return (
        <Paper className={classes.paper}>
          <Typography variant="display1" gutterBottom>
            Start messagingâ€¦
          </Typography>
          <Typography variant="body1" gutterBottom>
            Use <strong>Global</strong> to explore communities around here.
          </Typography>
          <Typography variant="body1" gutterBottom>
            Use <strong>Recents</strong> to see your recent conversations.
          </Typography>
        </Paper>
      );
    }

    return messages && messages.length ? (
      <div
        className={classes.messagesWrapper}
        ref={(wrapper) => {
          this.messagesWrapper = wrapper;
        }}
      >
        {messages &&
          messages.map(message => (
            <MessageItem
              message={message}
              // eslint-disable-next-line
              key={message._id}
              activeUser={activeUser}
            />
          ))}
      </div>
    ) : (
      <Typography variant="display1">There is no messages yet...</Typography>
    );
  }
}
export default withRouter(withStyles(styles)(MessageList));
