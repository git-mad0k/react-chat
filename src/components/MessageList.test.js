/* eslint-env jest */
import React from 'react';
import ReactDOM from 'react-dom';
import { MemoryRouter, Route } from 'react-router-dom';
import renderer from 'react-test-renderer';
import MessageList from './MessageList';

jest.mock('moment', () => () => ({ fromNow: () => '2 days ago' }));

const mockProps = {
  messages: [
    {
      _id: '1',
      chatId: '12345',
      content: 'Hello, World!',
      sender: {
        firstName: 'test',
        lastName: 'test',
        username: 'test',
        _id: '12345',
      },
      statusMessage: true,
      createdAt: '2018-03-16T10:53:23.200Z',
      updatedAt: '2018-03-16T10:53:23.200Z',
    },
    {
      _id: '2',
      chatId: '12345',
      content: 'Hello, React!',
      sender: {
        firstName: 'test',
        lastName: 'test',
        username: 'test',
        _id: '12345',
      },
      statusMessage: true,
      createdAt: '2018-03-16T10:53:23.200Z',
      updatedAt: '2018-03-16T10:53:23.200Z',
    },
  ],
  activeUser: {
    _id: '12345',
    username: 'test',
    isMember: true,
    isCreator: true,
    isChatMember: true,
  },
};
describe('<MessageList />', () => {
  it('render without crashing', () => {
    const div = document.createElement('div');
    ReactDOM.render(
      <MemoryRouter initialEntries={['/chat']}>
        <Route path="/chat/:chatId?" render={props => <MessageList {...mockProps} {...props} />} />
      </MemoryRouter>,
      div,
    );
    ReactDOM.unmountComponentAtNode(div);
  });

  it('reders correcty', () => {
    const tree = renderer
      .create(<MemoryRouter initialEntries={['/chat']}>
        <Route
          path="/chat/:chatId?"
          render={props => <MessageList {...mockProps} {...props} />}
        />
        {/* eslint-disable-next-line */}
        </MemoryRouter>,)
      .toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('renders messages', () => {
    const tree = renderer
      .create(<MemoryRouter initialEntries={['/chat/12345']}>
        <Route
          path="/chat/:chatId?"
          render={props => <MessageList {...mockProps} {...props} />}
        />
        {/* eslint-disable-next-line */}
        </MemoryRouter>,)
      .toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('renders without messages', () => {
    const tree = renderer
      .create(<MemoryRouter initialEntries={['/chat/12345']}>
        <Route
          path="/chat/:chatId?"
          render={props => <MessageList {...mockProps} {...props} messages={[]} />}
        />
        {/* eslint-disable-next-line */}
        </MemoryRouter>,)
      .toJSON();

    expect(tree).toMatchSnapshot();
  });
});
