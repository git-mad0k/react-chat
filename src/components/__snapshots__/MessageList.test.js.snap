// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<MessageList /> reders correcty 1`] = `
<div
  className="MuiPaper-root-58 MuiPaper-elevation2-62 MuiPaper-rounded-59 MessageList-paper-57"
>
  <h1
    className="MuiTypography-root-85 MuiTypography-display1-89 MuiTypography-gutterBottom-103"
  >
    Start messagingâ€¦
  </h1>
  <p
    className="MuiTypography-root-85 MuiTypography-body1-94 MuiTypography-gutterBottom-103"
  >
    Use 
    <strong>
      Global
    </strong>
     to explore communities around here.
  </p>
  <p
    className="MuiTypography-root-85 MuiTypography-body1-94 MuiTypography-gutterBottom-103"
  >
    Use 
    <strong>
      Recents
    </strong>
     to see your recent conversations.
  </p>
</div>
`;

exports[`<MessageList /> renders messages 1`] = `
<div
  className="MessageList-messagesWrapper-56"
>
  <div
    className="MessageItem-userStatusWrapper-118"
  >
    <div
      className="MessageItem-userName-119"
    >
      test
       
      <span
        className="MessageItem-userStatus-120"
      >
        Hello, World!
      </span>
    </div>
    <div
      className="MessageItem-userDateVisit-121"
    >
       
      2 days ago
       
    </div>
  </div>
  <div
    className="MessageItem-userStatusWrapper-118"
  >
    <div
      className="MessageItem-userName-119"
    >
      test
       
      <span
        className="MessageItem-userStatus-120"
      >
        Hello, React!
      </span>
    </div>
    <div
      className="MessageItem-userDateVisit-121"
    >
       
      2 days ago
       
    </div>
  </div>
</div>
`;

exports[`<MessageList /> renders without messages 1`] = `
<h1
  className="MuiTypography-root-85 MuiTypography-display1-89"
>
  There is no messages yet...
</h1>
`;
